README.md
[![#](https://img.shields.io/npm/v/@leiratech/trackzero-js)](https://www.npmjs.com/package/@leiratech/trackzero-js) ![#](https://img.shields.io/npm/l/@leiratech/trackzero-js)

<p align="center"><img alt="logo" src="./logo.svg" width="100"/></p>
<h1 align="center">
  TrackZero-JS
</h1>
Powerful, insightful and real-time analytics that helps you take your products and apps to the next level.

For more details, please visit [https://TrackZero.io](https://trackzero.io)

# Installation

```bat
npm install @leiratech/trackzero-js
```

# Setup

### Import

> Note: You can either use `import` or `require`, both work fine

```js
import { TrackZeroClient } from "@leiratech/trackzero-js";
```

### Initialization

Initialize the TrackZeroClient instance by passing in your API key before using any of the methods.

```js
let instance = new TrackZeroClient("API-KEY");
```

# Getting Started

### TrackZero Instance

After initializing the TrackZero instance, you could get the instance anywhere in your project

```js
let instance = TrackZeroClient.getInstance();
```

## Entity

Entities are objects that contain data. It can represent any object in the real world. Each entity is defined by it’s Type and Id.

```js
import { Entity } from "@leiratech/trackzero-js";
```

#### Create an Entity object

```js
/**
 * @constructor
 * Initializes the Entity object
 *
 * @param {string} type - states what the entity identifies as
 * @param {(string|number)} id - unique id associated with the entity
 */
let entity = new Entity("type", "id");
```

#### Add Attributes

```js
/**
 * Adds custom attributes to the entity
 *
 * @param {string} attribute - property name
 * @param {*} value - property value
 * @returns the entity instance
 */
entity.addAttribute("attribute", "value");
```

#### Add Attributes Referencing Other Entities

```js
/**
 * Adds custom attributes to the entity that are related to another entity
 *
 * @param {string} attribute - property name
 * @param {string} referenceType - the entity type it is referencing
 * @param {(string|number)} referenceId - the entity id it is referencing
 * @returns the entity instance
 */
entity.addEntityReferencedAttribute(
  "attribute",
  "referenceType",
  "referenceId"
);
```

#### Track Entity

```js
/**
 * Creates/Updates the Entity
 *
 * @async
 * @param {Entity} entity
 * @returns the response status
 */
await instance.upsertEntity(entity);
```

> **Note:** Upsert _(Update/Insert)_ is applied when sending the entity. So, if the the entity of type X with id Y already exists, then it gets updated, else a new entity is created. This also applies to the entity's referenced attributes in `addEntityReferencedAttribute`.

#### Complete Entity Example

```js
let user = new Entity("User", "USER_ID")
  .addAttribute("Name", "Sam Smith")
  .addAttribute("Date Of Birth", new Date(Date.UTC(1990, 11, 23))) //Make sure dates are in UTC
  .addEntityReferencedAttribute("Location", "Country", "US");

await instance.upsertEntity(user);
```

#### Delete Entity

> :warning: &nbsp; Deleting an entity will delete all events emitted by this entity.

> :exclamation: &nbsp; Deletion is permanent and cannot be undone.

```js
/**
 * Deletes the Entity
 *
 * @async
 * @param {string} type - type of entity to be deleted
 * @param {(string|number)} id - id of the entity to be deleted
 * @returns the response status
 */
await instance.deleteEntity("type", "id");
```

## Event

Events are considered as actions that are emitted from an entity and impact another entity

```js
import { Event } from "@leiratech/trackzero-js";
```

#### Create an Event object

```js
/**
 * @constructor
 * Initializes the Event object
 *
 * @param {string} emitterType - the entity type that emitted the event
 * @param {(string|number)} emitterId - the entity id that emitted the event
 * @param {string} name - the event name
 * ------Optional Params------
 * @param {(string|number)} [id] - the event id.
 * @param {string} [startTime] - (ISO 8601) The start time of the event.
 * @param {string} [endTime] - (ISO 8601) The end time of the event.
 */
let event = new Event(
  "emitterType",
  "emitterId",
  "name",
  "id",
  "startTime",
  "endTime"
);
```

> **Note:** Specifying your own event `id` prevents duplication if it happens and you send this event again. If not specified, it will be automatically set to a NewGuid.<br/>If `startTime` or `endTime` is not set, it will be automatically set to the current time UTC

#### Add Attributes

```js
/**
 * Adds custom attributes to the event
 *
 * @param {string} attribute - property name
 * @param {*} value - property value
 * @returns the event instance
 */
event.addAttribute("attribute", "value");
```

#### Add Attributes Referencing Other Entities

```js
/**
 * Adds custom attributes to the event that are related to another entity
 *
 * @param {string} attribute - property name
 * @param {string} referenceType - the entity type it is referencing
 * @param {(string|number)} referenceId - the entity id it is referencing
 * @returns the event instance
 */
event.addEntityReferencedAttribute("attribute", "referenceType", "referenceId");
```

#### Add Impacted Entities

```js
/**
 * Adds entities that were impacted by this event
 *
 * @param {string} impactedType - the entity type that was impacted by the event
 * @param {(string|number)} impactedId - the entity id that was impacted by the event
 * @returns the event instance
 */
event.addImpactedTarget("impactedType", "impactedId");
```

#### Track Event

```js
/**
 * Creates/Updates the Event
 *
 * @async
 * @param {Event} event
 * @returns the response status
 */
await instance.upsertEvent(event);
```

> **Note:** Like the Entity object, upsert _(Update/Insert)_ is applied to the event's emitter, entity's referenced attributes in `addEntityReferencedAttribute`, and the impacted entities in `addImpactedTarget`.

#### Complete Event Example

```js
let checked = new Event("User", "USER_ID", "Checked Out")
  .addAttribute("Cart Total", 99.95)
  .addEntityReferencedAttribute("Item", "Product", "SKU-1234")
  .addImpactedTarget("Warehouse", "WH-BLVD-652");

await instance.upsertEvent(checked);
```

#### Delete Event

> :exclamation: &nbsp; Deletion is permanent and cannot be undone.

```js
/**
 * Deletes the Event
 *
 * @async
 * @param {string} type - event type/name to be deleted
 * @param {(string|number)} id - id of the event to be deleted
 *
 * @returns the response status
 */
await instance.deleteEvent("type", "id");
```

# Resources

[Changelog](./CHANGELOG.md#change-log)

# License

[MIT](https://github.com/leiratech/TrackZero.JavaScript/blob/main/LICENSE)


CHANGELOG.MD
# Change Log

## v1.0.1

### Aug. 19, 2021

#### Bug Fixes

- Fixed issue when importing package in another project

#### Added

- [CHANGELOG.md](#change-log)
- Added getInstance static method to TrackZeroClient

#### Modifications

- Changed Endpoint to prevent Adbolck
- Validating methods' parameters
- JSDocs edited
- [README.md](./README.md) file updated
- Changed License to MIT

---

## v1.0.0

### Aug. 15, 2021

#### Description

Initial Release


network.js
const fetch = require("node-fetch");

const baseUrl = "https://api.trackzero.io/log";

const request = async function (endpointUrl, body, method) {
  try {
    const response = await fetch(`${baseUrl}${endpointUrl}`, {
      method,
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });

    let data = null;
    try {
      data = await response.json();
    } catch (error) {
      data = null;
    }

    return new Response(response.status, data);
  } catch (err) {
    return new Response(null, null, err);
  }
};

const api = {
  /**
   * @async
   * @param {Event|Entity} body
   * @returns response
   */
  post: async function (endpointUrl, body) {
    return await request(endpointUrl, body, "POST");
  },
  delete: async function (endpointUrl, body) {
    return await request(endpointUrl, body, "DELETE");
  },
};

class Response {
  constructor(status, data, error) {
    if (error) {
      this.error = `error: ${error}`;
    } else {
      this.status = status;
      this.data = data;
      this.error =
        status >= 400
          ? `Check status ${status} on https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses`
          : null;
    }
  }
}

module.exports = { api };


index.js
"use strict";
const { api } = require("./network");
/**
 * @class
 */
class TrackZeroClient {
  /**
   * @constructor
   * Initializes the singleton instance with the supplied API key
   *
   * @param {string} apiKey
   * @returns a singleton instance of TrackZeroClient
   */
  constructor(apiKey) {
    if (TrackZeroClient.instance instanceof TrackZeroClient) {
      return TrackZeroClient.instance;
    }

    if (!apiKey) throw "Error [TrackZeroClient][Constructor]: API Key Required";

    this.apiKey = encodeURI(apiKey);
    Object.freeze(this.apiKey);
    Object.freeze(this);

    TrackZeroClient.instance = this;
  }

  /**
   * Gets the singleton instance
   *
   * @static
   * @returns the instance of TrackZeroClient
   */
  static getInstance() {
    if (TrackZeroClient.instance instanceof TrackZeroClient) {
      return TrackZeroClient.instance;
    }
    throw "Error [TrackZeroClient][getInstance]: TrackZeroClient was not instantiated";
  }

  /**
   * Creates/Updates the Entity
   *
   * @async
   * @param {Entity} entity
   * @returns the response status
   */
  async upsertEntity(entity) {
    if (!(entity instanceof Entity)) {
      throw "Error [TrackZeroClient][upsertEntity]: parameter should be of type Entity";
    }
    let body = {
      type: entity.type,
      id: entity.id,
      customAttributes: entity.customAttributes,
    };

    return await api.post(`/entities?X-API-KEY=${this.apiKey}`, body);
  }

  /**
   * Deletes the Entity
   *
   * @async
   * @param {string} type - type of entity to be deleted
   * @param {(string|number)} id - id of the entity to be deleted
   * @returns the response status
   */
  async deleteEntity(type, id) {
    if (!type || !id) {
      throw "Error [Entity][deleteEntity]: 'type' and 'id' are required";
    }

    let body = { type, id };

    return await api.delete(`/entities?X-API-KEY=${this.apiKey}`, body);
  }

  /**
   * Creates/Updates the Event
   *
   * @async
   * @param {Event} event
   * @returns the response status
   */
  async upsertEvent(event) {
    if (!(event instanceof Event)) {
      throw "Error [TrackZeroClient][upsertEvent]: parameter should be of type Event";
    }
    let body = {
      id: event.id,
      emitter: {
        type: event.emitterType,
        id: event.emitterId,
      },
      startTime: event.startTime,
      endTime: event.endTime,
      name: event.name,
      CustomAttributes: event.customAttributes,
      Targets: event.targets,
    };

    return await api.post(`/events?X-API-KEY=${this.apiKey}`, body);
  }

  /**
   * Deletes the Event
   *
   * @async
   * @param {string} type - event type/name to be deleted
   * @param {(string|number)} id - id of the event to be deleted
   *
   * @returns the response status
   */
  async deleteEvent(type, id) {
    if (!type || !id) {
      throw "Error [Entity][deleteEvent]: 'type' and 'id' are required";
    }

    let body = { type, id };

    return await api.delete(`/events?X-API-KEY=${this.apiKey}`, body);
  }
}

/**
 * @class
 */
class Entity {
  /**
   * @constructor
   * Initializes the Entity object
   * @param {string} type - Entity's type
   * @param {(string|number)} id - Entity's id
   *
   * @example
   *
   * ```js
   * let entity = new Entity("User", "87717c11-ed5a...")
   *  //Adds an attribute to the entity (Method Chainable)
   *  .addAttribute("Name", "Sam Smith")
   *  //Adds an attribute that references another entity (Method Chainable)
   *  .addEntityReferencedAttribute("Nationality", "Country", "US")
   * ```
   */
  constructor(type, id) {
    if (!type || !id) {
      throw "Error [Entity][Constructor]: 'type' and 'id' are required";
    }
    this.type = type;
    this.id = id;
    this.customAttributes = {};
  }

  /**
   * Adds custom attributes to the entity
   *
   * @param {string} attribute - Entity's attribute name
   * @param {*} value - Entity's attribute value
   * @returns the entity instance
   */
  addAttribute(attribute, value) {
    if (!attribute || !value) {
      throw "Error [Entity][AddAttribute]: 'name' and 'value' are required";
    }
    this.customAttributes[attribute] = value;
    return this;
  }

  /**
   * Adds custom attributes to the entity that are related to another entity
   *
   * @param {string} attribute - Entity's attribute name
   * @param {string} referenceType - Related to entity type
   * @param {(string|number)} referenceId - Related to entity id
   * @returns the entity instance
   */
  addEntityReferencedAttribute(attribute, referenceType, referenceId) {
    if (!attribute || !referenceType || !referenceId) {
      throw "Error [Entity][AddEntityReferencedAttribute]: 'name', 'reference type' and 'reference id' are required";
    }
    this.customAttributes[attribute] = { type: referenceType, id: referenceId };
    return this;
  }
}

/**
 * @class
 */
class Event {
  /**
   * @constructor
   * Initializes the Event object
   *
   * @param {string} emitterType - Type of the entity that emitted the event.
   * @param {(string|number)} emitterId - Id of the entity that emitted the event.
   * @param {string} name - Event's name.
   * @param {(string|number)} [id] - Event's id. Specifying your own Id prevents duplication if it happens and you send this event again. When not specified, it will be automatically set to a NewGuid.
   * @param {string} [startTime] - (ISO 8601) The start time of the event. If not set, it will be automatically set to the current time UTC
   * @param {string} [endTime] - (ISO 8601) The end time of the event. If not set, it will be automatically set to the current time UTC
   *
   * @example
   *
   * ```js
   * let event = new Event("User", "87717c11-ed5a...", "Checked Out")
   *  //To add an attribute to the event (Method Chainable)
   *  .addAttribute("Cart Total", 799.99)
   *  //To add an attribute that references another entity (Method Chainable)
   *  .addEntityReferencedAttribute("Item", "Product", "SKU-1234")
   *  //To add an entity that was impacted by the event (Method Chainable)
   *  .addImpactedTarget("Company", "Store A")
   * ```
   */
  constructor(emitterType, emitterId, name, id, startTime, endTime) {
    if (!emitterType || !emitterId || !name) {
      throw "Error [Event][Constructor]: 'emitter type', 'emitter id' and 'name' are required";
    }
    this.name = name;
    this.emitterType = emitterType;
    this.emitterId = emitterId;
    this.id = id;
    this.startTime = startTime;
    this.endTime = endTime;
    this.targets = [];
    this.customAttributes = {};
  }

  /**
   * Adds custom attributes to the event
   *
   * @param {string} attribute - Event's attribute name
   * @param {*} value - Event's attribute value
   * @returns the event instance
   */
  addAttribute(attribute, value) {
    if (!attribute || !value) {
      throw "Error [Event][AddAttribute]: 'name' and 'value' are required";
    }
    this.customAttributes[attribute] = value;
    return this;
  }

  /**
   * Adds custom attributes to the event that are related to another entity
   *
   * @param {string} attribute - Event's attribute name
   * @param {string} referenceType - Related to entity type
   * @param {(string|number)} referenceId - Related to entity id
   * @returns the event instance
   */
  addEntityReferencedAttribute(attribute, referenceType, referenceId) {
    if (!attribute || !referenceType || !referenceId) {
      throw "Error [Event][AddEntityReferencedAttribute]: 'name', 'reference type' and 'reference id' are required";
    }
    this.customAttributes[attribute] = { type: referenceType, id: referenceId };
    return this;
  }

  /**
   * Adds entities that were impacted by this event
   *
   * @param {string} impactedType - Entity type that was impacted
   * @param {(string|number)} impactedId - Entity Id that was impacted
   * @returns the event instance
   */
  addImpactedTarget(impactedType, impactedId) {
    if (!impactedType || !impactedId) {
      throw "Error [Event][AddImpactedTarget]: 'type' and 'id' are required";
    }
    this.targets.push({ type: impactedType, id: impactedId });
    return this;
  }
}

module.exports = {
  Event,
  Entity,
  TrackZeroClient,
};
